packages/transport/src/transports/usb.ts

```diff
@@ -0,0 +1,353 @@
+import { createDeferred, AbortablePromise } from '@trezor/utils';
+
+import { Transport, AcquireInput } from './abstract';
+import { buildAndSend } from '../lowlevel/send';
+import { receiveAndParse } from '../lowlevel/receive';
+import { SessionsClient } from '../sessions/client';
+import { MessageFromTrezor, Descriptor, Success, ErrorGeneric } from '../types';
+
+import * as TRANSPORT_ERRORS from './errors';
+import * as COMMON_ERRORS from '../errors';
+import * as INTERFACE_ERRORS from '../interfaces/errors';
+import * as SESSION_ERRORS from '../sessions/errors';
+
+import type { TransportUsbInterface } from '../interfaces/usb';
+
+type UsbTransportConstructorParams = ConstructorParameters<typeof Transport>[0] & {
+    usbInterface: TransportUsbInterface;
+    sessionsClient: (typeof SessionsClient)['prototype'];
+};
+
+/**
+ *
+ * Abstract class for transports with direct usb access (webusb, nodeusb).
+ *
+ */
+export abstract class UsbTransport extends Transport {
+    // sessions client is a standardized interface for communicating with sessions backend
+    // which can live in couple of context (shared worker, local module, websocket server etc)
+    private sessionsClient: UsbTransportConstructorParams['sessionsClient'];
+    private transportInterface: TransportUsbInterface;
+
+    constructor({ messages, usbInterface, sessionsClient, signal }: UsbTransportConstructorParams) {
+        super({ messages, signal });
+        this.sessionsClient = sessionsClient;
+        this.transportInterface = usbInterface;
+    }
+
+    public async init() {
+        const handshakeRes = await this.sessionsClient.handshake();
+
+        if (!handshakeRes.success) {
+            return this.error({ error: TRANSPORT_ERRORS.SESSIONS_BACKGROUND_NOT_AVAILABLE });
+        }
+
+        return this.success(undefined);
+    }
+
+    public listen() {
+        if (this.listening) {
+            return this.error({ error: TRANSPORT_ERRORS.ALREADY_LISTENING });
+        }
+
+        this.listening = true;
+
+        // 1. transport interface reports descriptors change
+        this.transportInterface.on('transport-interface-change', devices => {
+            // 2. we signal this to sessions background
+            this.sessionsClient.enumerateDone({
+                paths: devices.map(d => d.path),
+            });
+        });
+        // 3. based on 2.sessions background distributes information about descriptors change to all clients
+        this.sessionsClient.on('descriptors', descriptors => {
+            // 4. we propagate new descriptors to higher levels
+            this.handleDescriptorsChange(descriptors);
+        });
+
+        return this.success(undefined);
+    }
+
+    public enumerate() {
+        return new AbortablePromise<
```

Comment:
use scheduledAction

Reply:
1dc9a648900cc6d6d102206286a90887f8fe85ce and a0022aa33f924d6d574ac28ddb91a31738ab578a